### PR TITLE
[#4205] Correctly set EPOLLOUT flag whe writeBytes(...) was not able …

### PR DESCRIPTION
**Pull request by normanmaurer**
…to write everything

Motivation:

writeBytes(...) missed to set EPOLLOUT flag when not all bytes were written. This could lead to have the EpollEventLoop not try to flush the remaining bytes once the socket becomes writable again.

Modifications:

- Move setting EPOLLOUT flag logic to one point so we are sure we always do it.
- Move OP_WRITE flag logic to one point as well.

Result:

Correctly try to write pending data if socket becomes writable again.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin please check

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java [hunk 1]
```diff
@@ -253,7 +251,6 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
                 if (done) {
                     in.remove();
                 } else {
-                    incompleteWrite(setOpWrite);
                     break;
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: `return;`?

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java [hunk 2]
```diff
@@ -219,13 +219,11 @@ protected void doWrite(ChannelOutboundBuffer in) throws Exception {
...
```
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: `return;`?

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java (hunk 1 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: @Scottmitch no we want break here as we want to have it trigger incompleteWrite(...) outside of the for loop in https://github.com/netty/netty/pull/4216/files#diff-1b9567b1b2c19df64161d297624b02a0R261

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java (hunk 2 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: @Scottmitch no we want break here as we want to have it trigger incompleteWrite(...) outside of the for loop in https://github.com/netty/netty/pull/4216/files#diff-1b9567b1b2c19df64161d297624b02a0R261

Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: Yip you are right...I remember now only need the first `return` :)

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java (hunk 1 above)
Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: Disregard

File: transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java (hunk 2 above)
Comment (normanmaurer) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: @Scottmitch yep.. that said let me add a comment why we break here and return before ... Just so we remember later on ;)

Comment (Scottmitch) [inline transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java]: @normanmaurer - Sure that is fine...more of just me rushing this time around. I got it the first time :)

**Comment by ninja-** [comment]
@normanmaurer ok so the $Entry problem can be separated into two parts. I am sure that the first part is fixed with a quick one-minute test, with this patch. I will test the rest on production and I am really expecting a drop in $Entry count from 5 000 000

**Comment by ninja-** [comment]
@normanmaurer you can pull(please bump 4.1 beta). If - assuming the application doesn't batch packets and every time calls writeAndFlush it would keep growing to millions, that would need a separate investigation anyway.

**Comment by Scottmitch** [comment]
@ninja- - This does not affect to account for back pressure if that is your question. Just ensure that we write what we have queued when the channel becomes writable.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (7961138f5278380da0397df5efb2863dad0b2333), 4.1 (076d4ed5147f186b5e5a25945b77bce63c0d0165) and master (943db1cea0b96208c060a675bc19b4031bfdd6b6)

**Pull request closed without merging**